### DISCUSSION TITLE
Upload/Download progress stream when using GridFS?

### DISCUSSION THREAD

**Discussion by JohnF17** [mongo-dart/mongo_dart#346] [Enhancements]
**Hello**, this probably is a feature request if it has **not been implemented yet** but I was just wondering if there was any implementation done for getting a stream for upload and download progress when using GridFS's **save** and/or **writeToFilename** method.

## Use Case
When uploading/downloading large chunks of data, the user would't be stuck there just watching the loading indicator to finish, they'll have an idea of how long it might take.

So If retrieving the progress is not done yet, here's an idea of how it might look (dio package inspired)

## For Uploads
```
var stream = dataFile.openRead();
var gridIn = gridFS.createFile(stream, courseFile.title);
await gridIn.save();
```

**The save method would look something like this** 

```
StreamController<double> streamController = StreamController();
var stream = dataFile.openRead();
var gridIn = gridFS.createFile(stream, filename);

await gridIn.save(
    onSendProgress: (count, total) {
        streamController.add(count / total);
    },
).then((Response response) {
    // Do something with the response 
})
.catchError((ex) {
    /// Show error dialog
    streamController.close();
}).whenComplete(() {
    streamController.close();
});

yield* streamController.stream;
```
## For Downloads
```
var gridOut = await gridFS.getFile(filename);
await gridOut.writeToFilename(result);
```

**The writeToFilename would look something like this** (_technically not sure if its the **writeToFilename** method that takes long or the **getFile** method but nonetheless;_)

```
StreamController<double> streamController = StreamController();
var gridOut = await gridFS.getFile(filename);

gridOut.writeToFilename(result,
    onReceiveProgress: (count, total) {
        streamController.add(count / total);
    },
).then((Response response) {
    // Do something with the response 
})
.catchError((ex) {
    /// Show error dialog
    streamController.close();
}).whenComplete(() {
    streamController.close();
});

yield* streamController.stream;
```

Ps. _Thanks for this awesome package_

**Comment by giorgiofran** [comment]
I'm reviewing the whole driver, so, at present, I have not the time to do this. I keep it highlighted as a new feature for when I will analyze also how to "restyle" the GridFS part.
